hasura/hasura-cloud-preview-apps
# feat: revoke connections before dropping ephemeral databases

## Description

### Problem

An ephemeral database gets created for each preview app setup on a pull request. The expected behavior is this should get dropped when the preview app shuts down after the PR is merged. Currently, dropping the ephemeral database fails as there may be active connections to the same. These connections should be revoked along with disabling any new connections before dropping the database.

### Solution

Solution reference: https://dba.stackexchange.com/a/11895

(The solution slightly differs for different versions of postgres.)

#### The solution is to run the following three additional queries:

- Check PG Version
  ```SQL
    SELECT VERSION();
  ```

- Revoke existing connections
  ```SQL
  SELECT pg_terminate_backend(pg_stat_activity.${pgStatActivityField})
  FROM pg_stat_activity
  WHERE pg_stat_activity.datname = '${dbName}'
  AND ${pgStatActivityField} <> pg_backend_pid();
  ```
  where pgStatActivityField = '`procpid`' for Postgres versions <=9.1 and '`pid`' for the versions after. 

- Disable new Connections
  ```SQL
  UPDATE pg_database SET datallowconn = 'false' WHERE datname = '${dbName}';
  ```
  This disables any new connections to the database for users and superusers alike.

- Finally, drop the database. The drop database query has to be coupled with the disable new connections query since creating a new `node-postgres` client to the database won't be possible after new connections are disabled.

## Steps to test

The testing was done by spinning up a local Postgres database making multiple connections to it and executing the preview apps flow.

- Spin up a local postgres instance
  ```bash
  docker run --rm -p 5432:5432 -e POSTGRES_PASSWORD=secret postgres
  ```
  To test for lower version (for eg. 9):
   ```bash
  docker run --rm -p 5432:5432 -e POSTGRES_PASSWORD=secret postgres@9
  ```
- Expose the port 5432 (Postgres database instance port) through ngrok tcp.
- Setup a Preview App repository, referencing the docs. You can refer to [my repository](https://github.com/OjasWadhwani/test-preview-apps) for testing. Make sure this has config workflows present (both create and delete) along with the metadata and migrations setup in the hasura folder.

  - Sample `hasura-cloud-preview-app.yaml` should look like this:
  
    ```yaml
    name: 'preview-apps'
    on: # rebuild any PRs and main branch changes
      pull_request:
    jobs:
      hasura-cloud-preview:
        runs-on: ubuntu-latest
        steps:
          - uses: actions/checkout@v2
          - uses: hasura/hasura-cloud-preview-apps@issue-73
            with:
              name: "test-preview-apps-${{github.env.GITHUB_HEAD_REF}}${{github.event.number}}"
              postgresDBConfig: |
                POSTGRES_SERVER_CONNECTION_URI=${{secrets.POSTGRES_SERVER_CONNECTION_URI}}
                PG_ENV_VARS_FOR_HASURA=PG_DATABASE_URL,PG_DB_URL_1,PG_DB_URL_2
              hasuraProjectDirectoryPath: hasura
              region: us-west-1
              tier: cloud_free
              hasuraEnv: |
                PG_DATABASE_URL=${{secrets.POSTGRES_SERVER_CONNECTION_URI}}
            env:
              GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
              HASURA_CLOUD_ACCESS_TOKEN: ${{secrets.HASURA_CLOUD_ACCESS_TOKEN}}
          - uses: hasura/comment-progress@v2.1.0
            with:
              github-token: ${{secrets.GITHUB_TOKEN}}
              id: preview_app_comment
              number: ${{github.event.number}}
              repository: "OjasWadhwani/test-preview-apps"
              message: |
                Console URL available at ${{steps.hasura_cloud_preview.outputs.consoleURL}}
                GraphQL Endpoint available at ${{steps.hasura_cloud_preview.outputs.graphQLEndpoint}}
    ````

    - Few things to note here: 
      - The repository version is @issue-73 (This branch version of the Hasura Preview Apps)
      - Take a note of the name format (test-preview-apps-<...>). This format should be the same in `delete-hasura-cloud-preview-app.yaml` as well.
      - `PG_DATABASE_URL` should be added under `PG_ENV_VARS_FOR_HASURA`
      - Connect to your test repo under "repository"
   
  - Sample `delete-hasura-cloud-preview-app.yaml` should look like:
    ```yaml
    on:
      pull_request:
        types: [closed]
    
    jobs:
      delete:
        runs-on: ubuntu-latest
        steps:
          - uses: actions/checkout@v2
    
          - name: Hasura Cloud Preview Apps
            uses: hasura/hasura-cloud-preview-apps@issue-73
            with:
              name: "test-preview-apps-${{github.env.GITHUB_HEAD_REF}}${{github.event.number}}"
              postgresDBConfig: |
                POSTGRES_SERVER_CONNECTION_URI=${{secrets.POSTGRES_SERVER_CONNECTION_URI}}
                PG_ENV_VARS_FOR_HASURA=PG_DATABASE_URL,PG_DB_URL_1,PG_DB_URL_2,PG_DB_URL3
              delete: true
            env:
              GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}} # ${{ secrets.GITHUB_TOKEN }} is provided by default by GitHub actions
              HASURA_CLOUD_ACCESS_TOKEN: ${{secrets.HASURA_CLOUD_ACCESS_TOKEN}} # Hasura Cloud access token to contact Hasura Cloud APIs
    ```
- Add the following secrets to your Github test repository settings:
  - `POSTGRES_SERVER_CONNECTION_URI` - The exposed ngrok URL wrapped under postgres connection string (Eg: `postgres://postgres:secret@8.tcp.ngrok.io:10076/)`
  - `HASURA_CLOUD_ACCESS_TOKEN` - THis needs to be generated as a Personal Access Token from your Hasura Cloud Settings page.
- Create a Pull Request to the test repository making the necessary changes to metadata/migrations. This should trigger the create preview app workflow and you should be able to see a project (with the previously mentioned naming format) on your Cloud dashboard after the workflow succeeds. 
- Make multiple connections to the ephemeral database:
  - This can be done in multiple ways:
    - Create a separate node-postgres client connecting to the database in a separate project:
      ```javascript
        const { Client } = require("pg");
        
        const client1 = new Client({
          connectionString: "postgres://postgres:secret@8.tcp.ngrok.io:10076/test_preview_apps_8",
        });
        
        client1.connect();
      ```
    - Alternatively, you can connect to the terminal through psql:
      ```bash
      psql postgres://postgres:secret@8.tcp.ngrok.io:10076/test_preview_apps_8
      ```
  - Note that the name of the databaseshould be the same as the one created on Cloud Dashboard.
- Merge the PR, and the preview app should successfully get deleted after all the connections should get revoked. If you set up the node-postgres client to test this, you should see the following error (expected since the connection was terminated from user's POV):
  ```bash
  error: terminating connection due to administrator command
  ```

## Limitation

A new Postgres Client has to be created for each query made. This is following the pattern of earlier implementations where the Postgres client ends after each function call and isn't reusable. Implementing a Pg [Pool](https://node-postgres.com/api/pool) could have been better but wasn't opted for given the time crunch and complexity of `CONNECTION_LIMIT` 

### Reference

Preview Apps docs: https://rehaul-preview-docs.hasura-docs-mono.pages.dev/docs/latest/graphql/cloud/deployment/preview-apps-github-action/
